In [1]:
#импрот библиотек 
import pandas as pd
import numpy as np

In [5]:
#чтение файла с помощью библиотеки pandas
df = pd.read_csv("6games.csv")

## Выводим первые 20 строк с помощью head:

In [6]:
df.head(20)

Name Platform  Year_of_Release  \
0                                     Wii Sports      Wii           2006.0   
1                              Super Mario Bros.      NES           1985.0   
2                                 Mario Kart Wii      Wii           2008.0   
3                              Wii Sports Resort      Wii           2009.0   
4                       Pokemon Red/Pokemon Blue       GB           1996.0   
5                                         Tetris       GB           1989.0   
6                          New Super Mario Bros.       DS           2006.0   
7                                       Wii Play      Wii           2006.0   
8                      New Super Mario Bros. Wii      Wii           2009.0   
9                                      Duck Hunt      NES           1984.0   
10                                    Nintendogs       DS           2005.0   
11                                 Mario Kart DS       DS           2005.0   
12                   Pokemon Gold/Pokemon Silver       GB           1999.0   
13                                       Wii Fit      Wii           2007.0   
14                            Kinect Adventures!     X360           2010.0   
15                                  Wii Fit Plus      Wii           2009.0   
16                            Grand Theft Auto V      PS3           2013.0   
17                 Grand Theft Auto: San Andreas      PS2           2004.0   
18                             Super Mario World     SNES           1990.0   
19  Brain Age: Train Your Brain in Minutes a Day       DS           2005.0   

           Genre  NA_sales  EU_sales  JP_sales  Other_sales  Critic_Score  \
0         Sports     41.36     28.96      3.77         8.45          76.0   
1       Platform     29.08      3.58      6.81         0.77           NaN   
2         Racing     15.68     12.76      3.79         3.29          82.0   
3         Sports     15.61     10.93      3.28         2.95          80.0   
4   Role-Playing     11.27      8.89     10.22         1.00           NaN   
5         Puzzle     23.20      2.26      4.22         0.58           NaN   
6       Platform     11.28      9.14      6.50         2.88          89.0   
7           Misc     13.96      9.18      2.93         2.84          58.0   
8       Platform     14.44      6.94      4.70         2.24          87.0   
9        Shooter     26.93      0.63      0.28         0.47           NaN   
10    Simulation      9.05     10.95      1.93         2.74           NaN   
11        Racing      9.71      7.47      4.13         1.90          91.0   
12  Role-Playing      9.00      6.18      7.20         0.71           NaN   
13        Sports      8.92      8.03      3.60         2.15          80.0   
14          Misc     15.00      4.89      0.24         1.69          61.0   
15        Sports      9.01      8.49      2.53         1.77          80.0   
16        Action      7.02      9.09      0.98         3.96          97.0   
17        Action      9.43      0.40      0.41        10.57          95.0   
18      Platform     12.78      3.75      3.54         0.55           NaN   
19          Misc      4.74      9.20      4.16         2.04          77.0   

   User_Score Rating  
0           8      E  
1         NaN    NaN  
2         8.3      E  
3           8      E  
4         NaN    NaN  
5         NaN    NaN  
6         8.5      E  
7         6.6      E  
8         8.4      E  
9         NaN    NaN  
10        NaN    NaN  
11        8.6      E  
12        NaN    NaN  
13        7.7      E  
14        6.3      E  
15        7.4      E  
16        8.2      M  
17          9      M  
18        NaN    NaN  
19        7.9      E

## Выполняем обзор данных:
### Предметная область - компьютерные игры.
Мы можем узнать название игры; платформу, на которой она была выпущена; год выпуска и жанр игры.

Также мы можем узнать продажи в определённых регионах, скорее всего это Северная Америка (NA), Европа (EU), Япония(JP) и других странах. Скорее всего продажи оцениваются в миллионах проданных копий. 

Кроме того из датасета можно узнать оценку критиков, оценку пользователей и какой-то общий рейтинг.



Датасет содержит следующие колонки: 

- Name — название игры;
- Platform — платформа;
- Year_of_Release — год выпуска;
- Genre — жанр игры;
- NA_sales — продажи в Северной Америке;
- EU_sales — продажи в Европе;
- JP_sales — продажи в Японии;
- Other_sales — продажи в других странах;
- Critic_Score — оценка критиков;
- User_Score — оценка пользователей;
- Rating — какой-то общий рейтинг. 

## Оценим данные: 

In [8]:
#оценка с помощью метода .info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16713 non-null  object 
 1   Platform         16715 non-null  object 
 2   Year_of_Release  16446 non-null  float64
 3   Genre            16713 non-null  object 
 4   NA_sales         16715 non-null  float64
 5   EU_sales         16715 non-null  float64
 6   JP_sales         16715 non-null  float64
 7   Other_sales      16715 non-null  float64
 8   Critic_Score     8137 non-null   float64
 9   User_Score       10014 non-null  object 
 10  Rating           9949 non-null   object 
dtypes: float64(6), object(5)
memory usage: 1.4+ MB


Этот метод возвращает название столбцов, типы данных, количество ненулевых объектов в каждом столбце.
- В датасете содержится 16715 строк и 11 столбцов.
- У 6 полей тип данных - float, у 5 полей тип данных - object.
- Столбцы названы некорректно (верхний регистр).
- В данных есть пропущенные значения, т.к. количество значений в столбцах различается.


## Выведем названия столбцов:

In [9]:
#вывод названий столбцов
df.columns

Index(['Name', 'Platform', 'Year_of_Release', 'Genre', 'NA_sales', 'EU_sales',
       'JP_sales', 'Other_sales', 'Critic_Score', 'User_Score', 'Rating'],
      dtype='object')

### Проблемы с названиями столбцов:
- Столбцы названы некорректно, все буквы должны быть в нижнем регистре.

Какие могли бы быть ещё проблемы: 
- Присутсвие пробелов в названии.
- Лишние символы в названии столбца.
- Отсутствие "змеиного стиля" (присутсвие нижнего подчеркивания) при названии из нескольких слов.
- В названиях используют разные языки (буквы легко спутать).
- Слишком непонятные названия столбцов, которые не отражают суть столбца.

In [10]:
# #приводим все наименования к нижнему регистру
df.columns = df.columns.str.lower()
df.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

## Поиск пропусков
Для наглядности создадим словарь с наименованием и описанием столбцов датафрейма и функцию с вызовом основных характеристик столбца:

In [11]:
# словарь с описанием столбцов
dict_df ={
     'name': 'название игры',
     'platform':'платформа',
     'year_of_release':'год выпуска',
     'genre': 'жанр игры',
     'na_sales': 'продажи в Северной Америке (миллионы проданных копий)',
     'eu_sales':'продажи в Европе (миллионы проданных копий)',
     'jp_sales': 'продажи в Японии (миллионы проданных копий)',
     'other_sales': 'продажи в других странах (миллионы проданных копий)',
     'critic_score': 'оценка критиков (максимум 100)', 
     'user_score': 'оценка пользователей (максимум 10)',
     'rating': 'рейтинг от организации ESRB (англ. Entertainment Software Rating Board)'
}

#словарь со значениями 
dict_name = pd.Series(dict_df)

In [16]:
#просмотр данных столбца: уникальные значения, тип данных, кол-во пропусков
def preprocessing(column):
    print(column, '-', dict_name[column])
    print('тип данных', df[column].dtypes)
    print('количество пропусков',df[column].isna().sum())
    if df[column].dtypes != 'object':
        print('минимальное значение:',df[column].min(), 'максимальное значение:', df[column].max() )
    print()
    print('уникальные значения:', '\n', df[column].value_counts())

### Столбцы name и genre:

In [17]:
preprocessing('name')

name - название игры
тип данных object
количество пропусков 2

уникальные значения: 
 Need for Speed: Most Wanted                         12
Ratatouille                                          9
LEGO Marvel Super Heroes                             9
FIFA 14                                              9
Madden NFL 07                                        9
                                                    ..
Jewels of the Tropical Lost Island                   1
Sherlock Holmes and the Mystery of Osborne House     1
The King of Fighters '95 (CD)                        1
Megamind: Mega Team Unite                            1
Haitaka no Psychedelica                              1
Name: name, Length: 11559, dtype: int64


In [18]:
preprocessing('genre')

genre - жанр игры
тип данных object
количество пропусков 2

уникальные значения: 
 Action          3369
Sports          2348
Misc            1750
Role-Playing    1498
Shooter         1323
Adventure       1303
Racing          1249
Platform         888
Simulation       873
Fighting         849
Strategy         683
Puzzle           580
Name: genre, dtype: int64


In [21]:
#посмотрим, что это за строки
df.loc[df['name'].isna()]

name platform  year_of_release genre  na_sales  eu_sales  jp_sales  \
659    NaN      GEN           1993.0   NaN      1.78      0.53      0.00   
14244  NaN      GEN           1993.0   NaN      0.00      0.00      0.03   

       other_sales  critic_score user_score rating  
659           0.08           NaN        NaN    NaN  
14244         0.00           NaN        NaN    NaN

В данных, которые содержат столбцы, одновременно 2 пропуска значений (нет названия игры и жанра) в строках 659 и 14244. Ввиду незначительности их количества строк удалим эти строки:

In [22]:
df.drop(index = [659, 14244], inplace = True)
df.loc[df['name'].isna()] #посмотрим, удалились ли строки

Empty DataFrame
Columns: [name, platform, year_of_release, genre, na_sales, eu_sales, jp_sales, other_sales, critic_score, user_score, rating]
Index: []

Вывод: удалены две строки с пропусками значений в обоих столбцах - name и genre.

### Столбец platform:

In [23]:
preprocessing('platform')

platform - платформа
тип данных object
количество пропусков 0

уникальные значения: 
 PS2     2161
DS      2151
PS3     1331
Wii     1320
X360    1262
PSP     1209
PS      1197
PC       974
XB       824
GBA      822
GC       556
3DS      520
PSV      430
PS4      392
N64      319
XOne     247
SNES     239
SAT      173
WiiU     147
2600     133
NES       98
GB        98
DC        52
GEN       27
NG        12
SCD        6
WS         6
3DO        3
TG16       2
GG         1
PCFX       1
Name: platform, dtype: int64


Вывод: нет явных проблем в данных столбца, пропусков также нет.

### Столбец year_of_release:

In [24]:
preprocessing('year_of_release')

year_of_release - год выпуска
тип данных float64
количество пропусков 269
минимальное значение: 1980.0 максимальное значение: 2016.0

уникальные значения: 
 2008.0    1427
2009.0    1426
2010.0    1255
2007.0    1197
2011.0    1136
2006.0    1006
2005.0     939
2002.0     829
2003.0     775
2004.0     762
2012.0     653
2015.0     606
2014.0     581
2013.0     544
2016.0     502
2001.0     482
1998.0     379
2000.0     350
1999.0     338
1997.0     289
1996.0     263
1995.0     219
1994.0     121
1993.0      60
1981.0      46
1992.0      43
1991.0      41
1982.0      36
1986.0      21
1989.0      17
1983.0      17
1990.0      16
1987.0      16
1988.0      15
1985.0      14
1984.0      14
1980.0       9
Name: year_of_release, dtype: int64


В столбце 269 пропусков, -1,61% от общей массы значений. Год релиза в дальнейшем исследовании имеет значение ,поэтому удалим их:

In [25]:
df.dropna(subset=['year_of_release'], inplace = True)
df.query('year_of_release.isna()') #посмотрим, удалились ли строки

Empty DataFrame
Columns: [name, platform, year_of_release, genre, na_sales, eu_sales, jp_sales, other_sales, critic_score, user_score, rating]
Index: []

Вывод: удалены строки с пропусками значений столбца year_of_release.

### Столбцы na_sales, eu_sales, jp_sales, other_sales:

In [26]:
preprocessing('na_sales')

na_sales - продажи в Северной Америке (миллионы проданных копий)
тип данных float64
количество пропусков 0
минимальное значение: 0.0 максимальное значение: 41.36

уникальные значения: 
 0.00     4449
0.02      562
0.03      554
0.04      543
0.05      536
         ... 
5.39        1
3.03        1
3.56        1
2.82        1
41.36       1
Name: na_sales, Length: 401, dtype: int64


In [27]:
preprocessing('eu_sales')

eu_sales - продажи в Европе (миллионы проданных копий)
тип данных float64
количество пропусков 0
минимальное значение: 0.0 максимальное значение: 28.96

уникальные значения: 
 0.00     5766
0.01     1467
0.02     1290
0.03      911
0.04      702
         ... 
1.52        1
5.75        1
2.17        1
2.78        1
28.96       1
Name: eu_sales, Length: 307, dtype: int64


In [28]:
preprocessing('jp_sales')

jp_sales - продажи в Японии (миллионы проданных копий)
тип данных float64
количество пропусков 0
минимальное значение: 0.0 максимальное значение: 10.22

уникальные значения: 
 0.00    10300
0.02      733
0.01      697
0.03      540
0.04      391
        ...  
3.14        1
2.73        1
3.18        1
3.63        1
0.97        1
Name: jp_sales, Length: 244, dtype: int64


In [29]:
preprocessing('other_sales')

other_sales - продажи в других странах (миллионы проданных копий)
тип данных float64
количество пропусков 0
минимальное значение: 0.0 максимальное значение: 10.57

уникальные значения: 
 0.00    6492
0.01    3398
0.02    1596
0.03     910
0.04     641
        ... 
2.11       1
7.53       1
1.21       1
1.79       1
0.93       1
Name: other_sales, Length: 155, dtype: int64


Вывод: в столбцах нет особенностей, на которые нужно обратить внимание на этапе предобработки.

### Столбец critic_score:

In [30]:
preprocessing('critic_score')

critic_score - оценка критиков (максимум 100)
тип данных float64
количество пропусков 8461
минимальное значение: 13.0 максимальное значение: 98.0

уникальные значения: 
 70.0    252
71.0    248
75.0    240
80.0    235
73.0    234
       ... 
20.0      3
29.0      3
21.0      1
17.0      1
13.0      1
Name: critic_score, Length: 81, dtype: int64


Вывод: пропуски значений столбца 'critic_score' оставлены без изменений во избежание искажения результатов корреляционного анализа и стат-тестов.

### Столбец user_score:

In [31]:
preprocessing('user_score')

user_score - оценка пользователей (максимум 10)
тип данных object
количество пропусков 6605

уникальные значения: 
 tbd    2376
7.8     322
8       285
8.2     276
8.3     252
       ... 
1.1       2
1.9       2
9.6       2
0         1
9.7       1
Name: user_score, Length: 96, dtype: int64


Поскольку значение tbd не означает наличие оценки пользователя, заменим ее на пропуск (nan), пропуски оставим незаполненными, чтобы впоследствии не исказить.

In [32]:
df.loc[df['user_score'] == 'tbd', 'user_score'] = float('nan')

Вывод: значения столбца 'user_score', равные tbd, заменены на nan (пропуски).Пропуски значений оставлены без изменений во избежание искажений результатов корреляционного анализа и стат-тестов.

### Столбец rating:

In [33]:
preprocessing('rating')

rating - рейтинг от организации ESRB (англ. Entertainment Software Rating Board)
тип данных object
количество пропусков 6676

уникальные значения: 
 E       3921
T       2905
M       1536
E10+    1393
EC         8
K-A        3
AO         1
RP         1
Name: rating, dtype: int64


Заменим пропуски на значение 'without rating':

In [34]:
df['rating'] = df['rating'].fillna('without rating')
preprocessing('rating')

rating - рейтинг от организации ESRB (англ. Entertainment Software Rating Board)
тип данных object
количество пропусков 0

уникальные значения: 
 without rating    6676
E                 3921
T                 2905
M                 1536
E10+              1393
EC                   8
K-A                  3
AO                   1
RP                   1
Name: rating, dtype: int64


Вывод: пропуски значений столбца rating заменены на значение 'without rating'.

## Проверка данных на дубликаты:

In [35]:
#проверка файлов на дубликаты:
df.duplicated().sum()

0

Дубикаты данных отсутствуют.

## Замена типа данных:
Меняем формат столбца year_of_release c float на int (поскольку нам нужно только значение года, смысла в преобразованиях в дату нет).

In [36]:
df['year_of_release'] = df['year_of_release'].astype(int)
df['year_of_release'].head()

0    2006
1    1985
2    2008
3    2009
4    1996
Name: year_of_release, dtype: int32

Заменим тип данных столбца user_score на float:

In [37]:
df['user_score'] = df['user_score'].astype(float)

Проверим внесение изменений:

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16444 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16444 non-null  object 
 1   platform         16444 non-null  object 
 2   year_of_release  16444 non-null  int32  
 3   genre            16444 non-null  object 
 4   na_sales         16444 non-null  float64
 5   eu_sales         16444 non-null  float64
 6   jp_sales         16444 non-null  float64
 7   other_sales      16444 non-null  float64
 8   critic_score     7983 non-null   float64
 9   user_score       7463 non-null   float64
 10  rating           16444 non-null  object 
dtypes: float64(6), int32(1), object(4)
memory usage: 1.4+ MB


## Создание сводной таблицы: 
Подсчитаем суммарные продажи во всех регионах.

In [51]:
df['total_sales'] = df['na_sales'] + df['eu_sales'] + df['jp_sales'] + df['other_sales']
df['total_sales'].head()

0    82.54
1    40.24
2    35.52
3    32.77
4    31.38
Name: total_sales, dtype: float64

Сгруппируем игры по платформам:

In [52]:
platform_year = df.groupby('platform')['total_sales'].agg(['count', 'sum']
                                    ).sort_values(by='sum', ascending = False).reset_index()

In [53]:
# топ-5 платформ по объемам продаж:
top_platforms = platform_year.head(5)['platform']
top_platforms

0     PS2
1    X360
2     PS3
3     Wii
4      DS
Name: platform, dtype: object

In [55]:
#сгруппируем топ платформ по годам релиза:
top = df.query('platform in @top_platforms').pivot_table(
                       index='year_of_release', columns = 'platform', 
                       values = 'total_sales', aggfunc = 'sum')
top

platform             DS     PS2     PS3     Wii    X360
year_of_release                                        
1985               0.02     NaN     NaN     NaN     NaN
2000                NaN   39.17     NaN     NaN     NaN
2001                NaN  166.43     NaN     NaN     NaN
2002                NaN  205.38     NaN     NaN     NaN
2003                NaN  184.31     NaN     NaN     NaN
2004              17.27  211.81     NaN     NaN     NaN
2005             130.14  160.66     NaN     NaN    8.25
2006             119.81  103.42   20.96  137.15   51.62
2007             146.94   75.99   73.19  152.77   95.41
2008             145.31   53.90  118.52  171.32  135.26
2009             119.54   26.40  130.93  206.97  120.29
2010              85.02    5.64  142.17  127.95  170.03
2011              26.18    0.45  156.78   59.65  143.84
2012              11.01     NaN  107.36   21.71   99.74
2013               1.54     NaN  113.25    8.59   88.58
2014                NaN     NaN   47.76    3.75   34.74
2015                NaN     NaN   16.82    1.14   11.96
2016                NaN     NaN    3.60    0.18    1.52

## Вывод:
В результате предобработки данных:
- Наименования стоблцов приведены к нижнему регистру.
- Удалены строки с пропусками данных в столбцах name и rating (2 строки), year_of_release (269 строк). Названия удалили, т.к. непоянтно, что там за игры и у них почти не было продаж. На общёю статистику удаление почти не влиет. Отсутствующие данные по оценкам и годам выпуска заменили на 0, в случае года такие данные восстановить невозможно, т.к. на разных платформах игра могла быть выпущена в разное время.
- Пропуски значений столбца rating (рэйтинг) заменены на значение 'without rating' (без рейтинга), т.к. у игр может быть различный рейтинг на различных платформах.
- Пропуски значений столбца user_score и critic_score оставлены без изменений во избежание искажений результатов корреляционного анализа и стат-тестов.
- Значения user_score (оценка пользователей), равные 'tbd' (To Be Determined) изменено на пропуски значений (значение nan).
- Подсчитаны и сохранены в столбце total_sales суммарные продажи во всех регионах, сгрупированы игры по платформам, вычеслены топ-5 платформ по объемам продаж и сгруппированы топ платформ по годам релиза с помощью pivot_table.
- Данные преобразовали к нужному типу. Год поменяли на целочисленный, так как по другому быть не может. Оценку пользователя привели к числу с плавающей запятой, так как там находятся дробные числа и в таблице есть аналогичный столбец с таким типом.


Данные готовы к анализу.